main.py

```diff
@@ -727,7 +733,129 @@ def visit_While(self, tree):
                 test, orelse))
 
     def visit_With(self, tree):
-        raise NotImplementedError('Open problem: with')
+
+        # Rewrite with-blocks as try-except statements as follows:
+        # with <expr> as <name>
+        #    <block>
+        # ----------------------
+        # __anonymous = <expr>
+        # bar = __anonymous.__enter__()
+        # try:
+        #     <block>
+        # except:
+        #     if __anonymous.__exit__(*sys.exc_info()):
+        #         raise
+        # else:
+        #     __anonymous.__exit__(None, None, None)
+        if tree.optional_vars is not None:
+            ctx_bind_name = tree.optional_vars.id
+        else:
+            ctx_bind_name = '__anonymous__enter__result'
+
+        ctx_manager_name = '__anonymous'
+
+        manager_assign = ast.Assign(
+            targets=[ast.Name(id=ctx_manager_name, ctx=ast.Store())],
+            value=tree.context_expr,
+        )
+        context_expr_assign = ast.Assign(
+            targets=[ast.Name(id=ctx_bind_name, ctx=ast.Store())],
+            value=ast.Call(
+                func=ast.Attribute(
+                    value=ast.Name(id=ctx_manager_name, ctx=ast.Store()),
+                    attr='__enter__',
+                    ctx=ast.Load(),
+                ),
+                args=[],
+                keywords=[],
+                starargs=None,
+                kwargs=None,
+            )
+        )
+        # Rewrite the with-block body as:
+        # try:
+        # <body>
+        # except:
+        #     if not <ctx_bind_name>.__exit__(*sys.exc_info()):
+        #         raise
+        sys_module = ast.Call(
+            func=ast.Name(id='__import__', ctx=ast.Load()),
+            args=[
+                ast.Str(s='sys'),
+            ],
+            keywords=[],
+            starargs=None,
+            kwargs=None,
+        )
+        none_node = ast.Name(id='None', ctx=ast.Load())
+
+        block = ast.TryExcept(
```

Comment:
Much of this is generated using an adapted version of the ast pretty printer I wrote [here](https://github.com/llllllllll/codetransformer/blob/master/codetransformer/utils/pretty.py).